### PR TITLE
Generate code using Go generics

### PR DESCRIPTION
**Pull request by vincentbernat**
The code mutation is a bit less pretty but it still works well. Tests
are kept in this case as they are easy to mutate to check that
everything is working.

However, this requires bumping the minimal Go version to 1.18.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -10,20 +10,39 @@ endif
...
+			$(SED) -i -E -e 's/\b'$$T'\b/\0[T]/g' *.go ; \
+			$(SED) -i -E -e 's/\b('$$T')\[T\]/\1[string]/g' *_test.go ; \
+		  done )
+	# Fix type definition
+	( cd generics_tree && $(SED) -i -E -e 's/^(type \w+)\[T\]/\1[T any]/' *.go)
+	# NewTreeVX function should be parametrized
+	( cd generics_tree && $(SED) -i -E -e 's/^(func NewTreeV.)/\1[T any]/' *.go)
+	( cd generics_tree && $(SED) -i -E -e 's/(NewTreeV.)\(/\1[string](/g' *_test.go)
+	# No need to cast interfaces
+	( cd generics_tree && $(SED) -i -E -e 's/\.\(string\)//g' *_test.go)
```
Comment (ingwarsw) [inline Makefile:39]: @vincentbernat Whats that part of makefile responsible for?
You have used is as part of generating `generics_tree` and it still can be used somehow?

**Comment by ingwarsw** [comment]
@vincentbernat Thanks for contribution.

Please give us some time to dig thru generics code and do internal testing.

File: Makefile (hunk 1 above)
Comment (vincentbernat) [reply Makefile:39]: I am unsure to understand your question. For `generics_tree`, the `codegen-%` rule is used as a first step, but further adaptations are needed. This code does that. The comments should show the different steps, specific to generics.

**Comment by vincentbernat** [comment]
I have pushed another commit to update the linter action to support Go 1.18. I think that's why it is failing. Also, before merging this PR, I suppose the README would need to be updated as well. Also, I wonder if there is any difference between code using generics and the specialized code. If not, an option would be to simply wait for Go 1.19 to be released, then Go 1.17 will be out of support and the code using generics could become the main one, without needing to generate stuff. Or it could be done in a v2 branch.

File: Makefile (hunk 1 above)
Comment (ingwarsw) [reply Makefile:39]: Ok.. now I understood.. thanks..

**Comment by vincentbernat** [comment]
I have updated the PR to make it work with PR #22.

**Review by i3149 — APPROVED** [review]
I think this looks good to me. At least it will not break anything.

**Comment by vincentbernat** [comment]
Note that it will prevent people using older versions of Go, even when they don't use generics.

**Comment by ingwarsw** [comment]
I just released `v1.1.0` that will be working with go 1.13+ and will release `v1.2.0` that will be compatible only with 1.18+
That whould protect people, and give place to upgrading tom latest compatible one..

**Pull request merged**